src/main/java/com/auth0/json/auth/TokenHolderDeserializer.java

```diff
@@ -0,0 +1,52 @@
+package com.auth0.json.auth;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Date;
+
+/**
+ * A custom Jackson Deserializer for the {@linkplain TokenHolder} object. It deserializes the JSON response and calculates
+ * an {@code expiresAt} date derived from the current time and the {@code expires_at} returned on the response.
+ */
+public class TokenHolderDeserializer extends StdDeserializer<TokenHolder> {
+
+    @SuppressWarnings("unused")
+    public TokenHolderDeserializer() {
+        this(null);
+    }
+
+    public TokenHolderDeserializer(Class<?> clazz) {
+        super(clazz);
+    }
+
+    @Override
+    public TokenHolder deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
+        JsonNode node = p.getCodec().readTree(p);
+
+        JsonNode accessTokenNode = node.get("access_token");
+        JsonNode idTokenNode = node.get("id_token");
+        JsonNode refreshTokenNode = node.get("refresh_token");
+        JsonNode tokenTypeNode = node.get("token_type");
+        JsonNode scopeNode = node.get("scope");
+        JsonNode expiresInNode = node.get("expires_in");
+
+        String accessToken = accessTokenNode != null ? accessTokenNode.asText() : null;
+        String idToken = idTokenNode !=  null ? idTokenNode.asText() : null;
+        String refreshToken = refreshTokenNode != null ? refreshTokenNode.asText() : null;
+        String tokenType = tokenTypeNode != null  ? tokenTypeNode.asText() : null;
+        String scope = scopeNode != null ?  scopeNode.asText() : null;
+
+        // As a primitive, if expires_in is not sent on the response, value will be 0 instead of null
+        long expiresIn = expiresInNode !=  null ? expiresInNode.asLong() : 0L;
```

Comment:
This preserves existing behavior of the default deserialization. It's not expected that we wouldn't receive an `expires_in`, but if we do, zero will be the value - and for consistency the `expiresAt` will be calculated using that.

Reply:
The `expires_in` property will only be present in the server's response if an Access Token was returned. If an ID token is all we get, this property will not be available. 

I think in that scenario it would be more accurate to set the "expires at" date to null instead of "now". 

Similarly, would be good to explore changing the `expires_in` type to an object instead, so it supports null values when the response doesn't include it. Thoughts?

Reply:
Changing the `expires_in` type to a boxed type would be a breaking change, as today its primitive value is zero if no access token was returned. If changed to a `Long` and set to `null` in this case, existing callers would receive a NPE if assigning the value to a primitive through the unboxing. So, we will leave it as a primitive.